File: problems/kissing-5d/compute/q4/t5_share.py
#!/usr/bin/env python3
"""36-cliques that share many vertices with a published 35.

q3 emptied share 34 and 33 (remove 1 add 2; remove 2 add 3).
q4 t5_repair_deep emptied share 32 and 31 (remove 3 add 4; remove 4 add 5).
A remaining 36-clique K satisfies |K ∩ C| ≤ 30 for every published
35-clique C, so it uses at least 6 outsiders, each adjacent to those
30 vertices of C.  Those outsiders therefore have d_C ≥ 30.

This file lists, for each published 35, the outsiders with d_C ≥ 30
and searches a 6-clique among them whose common neighbourhood in C
has size ≥ 30.  That is a complete search of 36-cliques that share
30 vertices with that C.  Lower shares are a further residue.
"""

from __future__ import annotations

import json
import sys
from itertools import combinations
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(ROOT))
sys.path.insert(0, str(ROOT / "q1"))
sys.path.insert(0, str(ROOT / "q2"))
sys.path.insert(0, str(ROOT / "q3"))

from t5_36 import bits_list, build_pool, is_clique


def clique_search(adj, n, target, node_limit=5_000_000):
    best = 0
    found = None
    nodes = 0

    def expand(P, stack):
        nonlocal best, found, nodes
        if found is not None:
            return
        nodes += 1
        if nodes > node_limit:
            return
        rsz = len(stack)
        if rsz + P.bit_count() <= best:
            return
        if P == 0:
            if rsz > best:
                best = rsz
            return
        rem = P
        ord_v, col = [], []
        c = 0
        while rem:
            c += 1
            avail = rem
            while avail:
                v = (avail & -avail).bit_length() - 1
                ord_v.append(v)
                col.append(c)
                avail &= ~adj[v]
                avail &= ~(1 << v)
                rem &= ~(1 << v)
        Q = P
        for i in range(len(ord_v) - 1, -1, -1):
            if found is not None or nodes > node_limit:
                return
            if rsz + col[i] <= best:
                return
            v = ord_v[i]
            stack.append(v)
            if rsz + 1 >= target:
                found = list(stack)
                best = rsz + 1
                return
            expand(Q & adj[v], stack)
            stack.pop()
            Q &= ~(1 << v)

    expand((1 << n) - 1, [])
    return found, best, nodes, found is not None or nodes <= node_limit


def main() -> int:
    G = build_pool()
    adj, n = G["adj"], G["n"]
    report = {"n": n, "by_code": {}, "found_36": False, "complete_share30": True}

    for name, rec in G["published"].items():
        C = rec["remainder_clique"]
        if len(C) != 35 or not is_clique(adj, C):
            continue
        Cset = set(C)
        Cbits = 0
        for v in C:
            Cbits |= 1 << v
        degC = []
        heavy = []
        for v in range(n):
            if v in Cset:
                continue
            d = bin(adj[v] & Cbits).count("1")
            degC.append(d)
            if d >= 30:
                heavy.append((v, d))
        hist = {}
        for d in degC:
            hist[d] = hist.get(d, 0) + 1
        print(f"{name} outside=320 heavy_dC>=30: {len(heavy)} "
              f"degC_hist={dict(sorted(hist.items()))}", flush=True)

        rec_out = {
            "n_heavy": len(heavy),
            "degC_hist": {str(a): b for a, b in sorted(hist.items())},
            "found": False,
            "best_outsiders": 0,
            "complete": True,
        }
        if len(heavy) >= 6:
            verts = [v for v, _ in heavy]
            remap = {v: i for i, v in enumerate(verts)}
            m = len(verts)
            hadj = [0] * m
            for i, a in enumerate(verts):
                for j, b in enumerate(verts):
                    if i < j and ((adj[a] >> b) & 1):
                        hadj[i] |= 1 << j
                        hadj[j] |= 1 << i
            # search 6-clique, then check common in C
            hit, best, nodes, ok = clique_search(hadj, m, 6)
            rec_out["best_outsiders"] = best
            rec_out["nodes"] = nodes
            rec_out["complete"] = ok
            print(f"  6-clique among heavy: best={best} found={hit is not None} "
                  f"nodes={nodes} complete={ok}", flush=True)
            if hit is not None:
                O = [verts[i] for i in hit]
                common = Cbits
                for v in O:
                    common &= adj[v]
                ncommon = bin(common).count("1")
                rec_out["n_common"] = ncommon
                if ncommon >= 30 and is_clique(adj, O + bits_list(common)[:30]):
                    K = O + bits_list(common)[:30]
                    rec_out["found"] = True
                    report["found_36"] = True
                    report["clique36"] = K[:36]
            # even if a 6-clique exists, filter by common neighbourhood
            # exhaustively when m is small
            if m <= 28 and not rec_out["found"]:
                hits = 0
                for add in combinations(range(m), 6):
                    if not is_clique(hadj, add):
                        continue
                    O = [verts[i] for i in add]
                    common = Cbits
                    for v in O:
                        common &= adj[v]
                    if bin(common).count("1") >= 30:
                        hits += 1
                        rec_out["found"] = True
                        report["found_36"] = True
                        report["clique36"] = O + bits_list(common)[:30]
                        break
                rec_out["six_cliques_checked"] = True
                rec_out["hits_share30"] = hits
        else:
            rec_out["best_outsiders"] = len(heavy)
            rec_out["complete"] = True
        report["by_code"][name] = rec_out
        report["complete_share30"] = report["complete_share30"] and rec_out["complete"]

    (HERE / "t5_share.json").write_text(json.dumps(
        {k: v for k, v in report.items() if k != "clique36"}, indent=2
    ) + "\n")
    print("found_36", report["found_36"],
          "complete_share30", report["complete_share30"])
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: problems/kissing-5d/compute/q4/t5_share.c
/* Exact 36-cliques that share s vertices with a published 35.
 *
 * Usage: ./t5_share <adj.txt> <published.json> <share> [node_limit]
 *
 * For each published 35-clique P and each (35-share)-subset L of P,
 * S = P \ L has size share.  A 36-clique meeting P in S is S plus a
 * (36-share)-clique in the common neighbourhood of S minus S.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#define MAXN 512
#define W ((MAXN + 63) / 64)

typedef uint64_t u64;

static int n, nwords, target;
static u64 adj[MAXN][W];
static int best, found;
static long nodes, node_limit;
static int found_idx[64];

static int popc(const u64 *a)
{
    int s = 0;
    for (int i = 0; i < nwords; i++)
        s += __builtin_popcountll(a[i]);
    return s;
}

static int first_bit(const u64 *a)
{
    for (int i = 0; i < nwords; i++)
        if (a[i])
            return i * 64 + __builtin_ctzll(a[i]);
    return -1;
}

static void bit_clear(u64 *a, int v)
{
    a[v >> 6] &= ~(1ULL << (v & 63));
}

static int bit_test(const u64 *a, int v)
{
    return (int)((a[v >> 6] >> (v & 63)) & 1U);
}

static void bit_set(u64 *a, int v)
{
    a[v >> 6] |= 1ULL << (v & 63);
}

static int colour_order(const u64 *P, int *ord, int *col)
{
    u64 rem[W];
    memcpy(rem, P, nwords * sizeof(u64));
    int m = 0, c = 0;
    while (popc(rem)) {
        c++;
        u64 avail[W];
        memcpy(avail, rem, nwords * sizeof(u64));
        int v;
        while ((v = first_bit(avail)) >= 0) {
            ord[m] = v;
            col[m] = c;
            m++;
            bit_clear(avail, v);
            for (int w = 0; w < nwords; w++)
                avail[w] &= ~adj[v][w];
            bit_clear(rem, v);
        }
    }
    return m;
}

static void expand(u64 *P, int rsz, int *stack)
{
    nodes++;
    if (found || nodes > node_limit)
        return;
    int psz = popc(P);
    if (rsz + psz <= best)
        return;
    if (psz == 0) {
        if (rsz > best)
            best = rsz;
        return;
    }
    int ord[MAXN], col[MAXN];
    int m = colour_order(P, ord, col);
    u64 Q[W];
    memcpy(Q, P, nwords * sizeof(u64));
    for (int i = m - 1; i >= 0; i--) {
        if (found || nodes > node_limit)
            return;
        if (rsz + col[i] <= best)
            return;
        int v = ord[i];
        u64 P2[W];
        for (int w = 0; w < nwords; w++)
            P2[w] = Q[w] & adj[v][w];
        stack[rsz] = v;
        if (rsz + 1 >= target) {
            found = 1;
            best = rsz + 1;
            memcpy(found_idx, stack, (size_t)target * sizeof(int));
            return;
        }
        expand(P2, rsz + 1, stack);
        bit_clear(Q, v);
    }
}

static int read_graph(FILE *fp)
{
    if (fscanf(fp, "%d", &n) != 1)
        return 0;
    if (n <= 0 || n > MAXN)
        return 0;
    nwords = (n + 63) / 64;
    memset(adj, 0, sizeof adj);
    for (int i = 0; i < n; i++) {
        for (int w = 0; w < nwords; w++) {
            unsigned long long x;
            if (fscanf(fp, "%llx", &x) != 1)
                return 0;
            adj[i][w] = (u64)x;
        }
        bit_clear(adj[i], i);
    }
    return 1;
}

static int parse_published(const char *path, int cl[4][40], int *ncl)
{
    FILE *f = fopen(path, "r");
    if (!f)
        return 0;
    char buf[1 << 16];
    size_t m = fread(buf, 1, sizeof buf - 1, f);
    buf[m] = 0;
    fclose(f);
    *ncl = 0;
    const char *p = buf;
    while (*p && *ncl < 4) {
        const char *r = strstr(p, "remainder_clique");
        if (!r)
            break;
        r = strchr(r, '[');
        if (!r)
            break;
        r++;
        int k = 0;
        while (*r && *r != ']') {
            if (*r >= '0' && *r <= '9') {
                int v = 0;
                while (*r >= '0' && *r <= '9') {
                    v = v * 10 + (*r - '0');
                    r++;
                }
                if (k < 40)
                    cl[*ncl][k++] = v;
            } else {
                r++;
            }
        }
        if (k == 35)
            (*ncl)++;
        p = r;
    }
    return *ncl == 4;
}

int main(int argc, char **argv)
{
    if (argc < 4) {
        fprintf(stderr, "usage: %s adj.txt published.json share [node_limit]\n",
                argv[0]);
        return 1;
    }
    FILE *fp = fopen(argv[1], "r");
    if (!fp || !read_graph(fp)) {
        fprintf(stderr, "bad graph\n");
        return 1;
    }
    fclose(fp);
    int cl[4][40], ncl;
    if (!parse_published(argv[2], cl, &ncl)) {
        fprintf(stderr, "bad published json (need 4 remainder 35-cliques)\n");
        return 1;
    }
    int share = atoi(argv[3]);
    node_limit = (argc >= 5) ? atol(argv[4]) : 200000L;
    if (share < 20 || share > 34) {
        fprintf(stderr, "share out of range\n");
        return 1;
    }
    int leave = 35 - share;
    int need = 36 - share;
    long tried = 0, empty = 0, incomplete = 0, hits = 0;
    int best_out = 0;
    int comb[16];

    printf("{\n  \"n\": %d,\n  \"share\": %d,\n  \"leave\": %d, \"need\": %d,\n",
           n, share, leave, need);
    printf("  \"by_code\": {\n");

    for (int c = 0; c < ncl; c++) {
        long c_tried = 0, c_empty = 0, c_inc = 0, c_hit = 0;
        int c_best = 0;
        /* enumerate leave-subsets of the 35 */
        for (int i = 0; i < leave; i++)
            comb[i] = i;
        for (;;) {
            u64 S[W];
            memset(S, 0, sizeof S);
            int inL[MAXN];
            memset(inL, 0, sizeof inL);
            for (int i = 0; i < leave; i++)
                inL[cl[c][comb[i]]] = 1;
            for (int i = 0; i < 35; i++)
                if (!inL[cl[c][i]])
                    bit_set(S, cl[c][i]);
            u64 N[W];
            memset(N, 0xFF, sizeof N);
            if (n & 63)
                N[nwords - 1] = (1ULL << (n & 63)) - 1;
            else
                N[nwords - 1] = ~0ULL;
            for (int i = 0; i < 35; i++) {
                int v = cl[c][i];
                if (inL[v])
                    continue;
                for (int w = 0; w < nwords; w++)
                    N[w] &= adj[v][w];
            }
            for (int w = 0; w < nwords; w++)
                N[w] &= ~S[w];
            target = need;
            found = 0;
            best = need - 1;
            if (c_best > best)
                best = c_best;
            nodes = 0;
            int stack[64];
            expand(N, 0, stack);
            c_tried++;
            tried++;
            if (best > c_best)
                c_best = best;
            if (best > best_out)
                best_out = best;
            if (found) {
                c_hit++;
                hits++;
            } else if (nodes > node_limit) {
                c_inc++;
                incomplete++;
            } else {
                c_empty++;
                empty++;
            }
            /* next combination */
            int i = leave - 1;
            while (i >= 0 && comb[i] == 35 - leave + i)
                i--;
            if (i < 0)
                break;
            comb[i]++;
            for (int j = i + 1; j < leave; j++)
                comb[j] = comb[j - 1] + 1;
        }
        if (c)
            printf(",\n");
        printf("    \"%d\": {\"tried\": %ld, \"empty\": %ld, \"incomplete\": %ld, "
               "\"found\": %ld, \"best_extra\": %d, \"complete\": %s}",
               c, c_tried, c_empty, c_inc, c_hit, c_best,
               c_inc == 0 && c_hit == 0 ? "true" : "false");
        fprintf(stderr, "code %d tried=%ld empty=%ld inc=%ld hit=%ld best=%d\n",
                c, c_tried, c_empty, c_inc, c_hit, c_best);
    }
    printf("\n  },\n");
    printf("  \"tried\": %ld,\n  \"empty\": %ld,\n  \"incomplete\": %ld,\n",
           tried, empty, incomplete);
    printf("  \"found_36\": %s,\n", hits ? "true" : "false");
    printf("  \"best_extra\": %d,\n", best_out);
    printf("  \"complete\": %s\n}\n",
           incomplete == 0 && hits == 0 ? "true" : "false");
    return 0;
}
